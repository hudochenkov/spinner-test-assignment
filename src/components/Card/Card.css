.card {
	margin: 50px auto;
	padding: 20px;
	width: 200px;
	border-radius: 10px;
	background: #fff;
	box-shadow: 0 10px 30px 0 rgba(0, 0, 0, 0.2);
}
.card__progress {
	display: flex;
	justify-content: center;
	margin-bottom: 30px;
}
.card__actions {
	margin: 0;
	text-align: center;
}
.card__action {
	padding: 0 35px;
	height: 40px;
	outline: none;
	border: none;
	border-radius: 20px;
	background: hsl(210, 100%, 63%);
	color: #fff;
	font-size: 16px;
	font-family: inherit;
	cursor: pointer;
}
.card__action:hover,
.card__action:focus {
	background: hsl(210, 100%, 53%);
}
.card__action:active {
	background: hsl(210, 100%, 43%);
}
