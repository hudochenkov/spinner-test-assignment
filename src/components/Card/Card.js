import React, { Component } from 'react';
import './Card.css';
import ProgressRing from '../ProgressRing';

// Default state here, because we're reusing in, when reseting a state
const defaultState = {
	inProgress: false,
	progress: 0,
};

class Card extends Component {
	constructor() {
		super();

		this.state = {
			...defaultState,
		};

		this.startProgress = this.startProgress.bind(this);
		this.stopProgress = this.stopProgress.bind(this);
	}

	startProgress() {
		this.setState({
			inProgress: true,
		});

		// Faking percentage change
		this.progressInterval = setInterval(() => {
			if (this.state.progress >= 100) {
				clearInterval(this.progressInterval);
			} else {
				this.setState(prevState => ({
					// Don't set value more than 100
					progress: Math.min(prevState.progress + 2, 100),
				}));
			}
		}, 100);
	}

	stopProgress() {
		clearInterval(this.progressInterval);

		this.setState({
			...defaultState,
		});
	}

	render() {
		return (
			<div className="card">
				<div className="card__progress">
					<ProgressRing value={this.state.progress} isActive={this.state.inProgress} />
				</div>

				<p className="card__actions">
					{this.state.inProgress ? (
						<button className="card__action" type="button" onClick={this.stopProgress}>
							Stop
						</button>
					) : (
						<button className="card__action" type="button" onClick={this.startProgress}>
							Start
						</button>
					)}
				</p>
			</div>
		);
	}
}

export default Card;
