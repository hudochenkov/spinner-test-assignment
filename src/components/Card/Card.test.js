import React from 'react';
import { shallow } from 'enzyme';
import Card from './Card';

describe('<Card />', () => {
	it('default', () => {
		const wrapper = shallow(<Card />);

		wrapper.find('.card__action').simulate('click');

		expect(wrapper).toMatchSnapshot();
	});

	it('actions', () => {
		const wrapper = shallow(<Card />);

		// Clicking Start
		wrapper.find('.card__action').simulate('click');
		expect(wrapper.state('inProgress')).toBe(true);
		expect(wrapper.find('.card__action').text()).toEqual('Stop');

		// Clicking Stop
		wrapper.find('.card__action').simulate('click');
		expect(wrapper.state('inProgress')).toBe(false);
		expect(wrapper.find('.card__action').text()).toEqual('Start');
	});
});
