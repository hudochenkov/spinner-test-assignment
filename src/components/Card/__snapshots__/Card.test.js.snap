// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> default 1`] = `
<div
  className="card"
>
  <div
    className="card__progress"
  >
    <ProgressRing
      isActive={true}
      value={0}
    />
  </div>
  <p
    className="card__actions"
  >
    <button
      className="card__action"
      onClick={[Function]}
      type="button"
    >
      Stop
    </button>
  </p>
</div>
`;
